docs: tenants (#67474)
metabase/metabase: docs/embedding/tenants.md

---
title: Tenants
---

# Tenants

{% include plans-blockquote.html feature="Tenants" is_plural=true %}

> Tenants simplify user provisioning and configuration, especially in embedding contexts.

Reviewer: Think we should say what tenants are first. And maybe emphasize the value proposition. The list already nicely explains why you'd use them.

```suggestion
A Tenant is a named set of attributes you can assign to a user to isolate them from other tenants. If you're building a SaaS app with embedded Metabase dashboards, you can assign the customers of your SaaS to tenants. 

The big advantage of grouping users into tenants is that you can use the same content and group permissions for all tenants, so you don't have to create all new stuff for each tenant, and each tenant only ever sees their own data.
```

You can use tenants to:

- simplify bulk permissions through tenant groups and attributes
- create collections of shared assets to avoid duplicating dashboards,
- and provision tenants with SSO.

To get started with tenants:

1. [Get familiar with tenant concepts](LINK)
2. [Enable tenants](LINK)
3. Create a [tenant](LINK) and [tenant users](LINK), or [provision them with SSO](LINK)
4. [Create shared collections](LINK)
5. Set up [data](LINK) and [collection](LINK) permissions

> You can also

Reviewer: ```suggestion
```

## Concepts

**Tenant** is an abstraction representing groups of users that share some properties but should be isolated from other users. For example, if you're building a SaaS app with embedded Metabase dashboards, the customers of your SaaS are tenants.

> While working with tenants in Metabase, you'll encounter different user and group types, and some some special collection types. Let's establish some terminology.

Reviewer: ```suggestion
While working with tenants in Metabase, you'll encounter different user and group types, and some special collection types. Let's establish some terminology.
```

### User types

> - **Tenant users** are the end users. In a B2B SaaS context, these are the your customer's user. Usually they will be interacting with Metabase through an intermediate app (for example, through an embedded dashboard).

Reviewer: ```suggestion
- **Tenant users** are the end users. In a B2B SaaS context, these are the your customer's users. Usually tenant users interact with Metabase through an intermediate app (for example, through an embedded dashboard).
```

> - **Internal users** are Metabase users who don't belong to any tenant. Think Metabase admins, or devs building the dashboards that will be shared with tenants.

Reviewer: ```suggestion
- **Internal users** are "normal" Metabase users who don't belong to any tenant. Think Metabase admins, or devs building the dashboards that will be shared with tenants.
```

### Group types

Tenant users and and internal users can be organized into Metabase [groups](LINK).

[SCREENSHOT]

- **All tenant users** is a special group representing all individual end users across all tenants.

>   This group can be used to [configure permissions](LINK) for tenant users. If you need more granular permission controls within each tenant, you can also created tenant groups.

Reviewer: ```suggestion
  This group can be used to [configure permissions](LINK) for all tenant users. If you need more granular permission controls within each tenant, you can also create tenant groups.
```

- **Tenant groups** can be used to create additional permission levels for tenant users.

>   For example, if you're making recruiting software, every end user can be a recruiter (with access to all data about all open roles) or a hiring manager (with access to analytics for a specific role). So in Metabase, you can create two tenant groups - "Recruiters" and "Hiring managers" and configure appropriate permissions.

Reviewer: I like the examples throughout.

2. Change the **Tenant assignment attribute** key to your preferred identifier.

### Provisioning tenants and users

You can [turn on JWT user provisioning](LINK) so that Metabase will automatically create users and tenants mentioned in the JWT.

When user provisioning with JWT is enabled:

1. Metabase reads the tenant identifier from the JWT claim. By default, this is the `@tenant` key (you can configure this)
2. If the tenant doesn't exist, Metabase automatically creates it. Metabase will use the value of the `@tenant` key (or your chosen assignemnt attribute) as the tenant slug.
3. New users are automatically assigned to the tenant from their JWT.

### Troubleshooting JWT authentication with tenants

Some auth common error messages and what they mean:

- **Cannot add tenant claim to internal user**: JWT includes a tenant, but the user is an internal user. Only tenant users can have a tenant.
- **Tenant claim required for external user**: JWT lacks a tenant claim, but the user is an external user.
- **Tenant ID mismatch with existing user**: JWT has a different tenant than the user's assigned tenant.
- **Tenant is not active**: The tenant exists but has been deactivated.

> ## Data permissions for tenants

Author: I'll keep it in first draft, while i work on a separate doc "Permissions for embedding" that will have this info and more.

### Use tenant attributes for data permissions

[Row and column security](LINK), [Impersonation](LINK), and Database routing require user attributes. You can [specify custom tenant attributes](LINK) to configure data permissions based on attribute values. See [Tenant attributes](LINK). Alternatively, you can use the special tenant slug attribute, see [Special slug attribute](LINK).

## Collection permissions for tenants

Collection permissions control which entities (dashboards, questions, models etc) people can see.

To configure what _data_ can people see in those entities, and what they can do with that data, see [Data permissions](LINK) instead.

In Metabase, there are 3 levels of collection permissions: **No** access, **View**-only access, and **Curate** access (allows for creating and saving new entities like dashboards). For more general information about collection permissions in Metabase, see [Collection permissions](LINK).

Permissions are granted to groups. Which permissions are available to each group depend on the type of the group (external/tenant or internal) and the type of the collection.

### Tenant user collection permissions

> - For **internal collection**, tenant users will have **No** access.

Reviewer: ```suggestion
- For **internal collections**, tenant users will have **No** access.
```

You can't even change this, right? It is law.

> - For **shared collections**, tenant users can only have **View** or **No** access. This means that at _most_, tenant users can can see existing entities but not create new ones.

Reviewer: ```suggestion
- For **shared collections**, tenant users can only have **View** or **No** access. This means that at _most_, tenant users can see existing entities but not create new ones.
```

- For their own **personal collections**, tenant users will always have **Curate** permissions.

### Internal user collection permissions

- Metabase Admins will have **Curate** access to all shared collections and all tenant collections.
- Other internal groups can be granted **View** or **Curate** access to **shared collections**, see [Configuring shared collection permissions](LINK),
- Non-admin internal users will have **No** access to tenant-specific collections. Currently, this can't be configured.

For configuring permissions to _internal_ collections for internal users, see [general docs on collection permissions](LINK)

### Configuring shared collections permissions

To configure access to shared collections for tenant and internal groups, go to **Admin settings > Permissions > Shared collections**.

You can configure access for each shared collections and their subcollections for both internal and external users. See general docs on [collection permissions](LINK).

> Special **Root shared collection** controls who has access to _all_ shared collections. For example, if you want to make sure you internal users don't have access to any tenant shared collections, you can revoke permissions from Root shared collection.

Reviewer: ```suggestion
Special **Root shared collection** controls who has access to _all_ shared collections. For example, if you want to make sure your internal users don't have access to any tenant shared collections, you can revoke permissions from the Root shared collection.
```

### From multi-tenant to single-tenant

If you disable multi-tenant strategy, _all your tenant users will be deactivated_ and _all collections will be deleted_ (although you'll get both users and collections back if you reactivate multi-tenant strategy later). So if you want to wish to keep the active users but just disable tenancy features, you'll need some extra prep.

1. Replicate the tenant setup you have with regular groups and collections (instead of tenant groups and shared collections). Review documentation for Data permissions, Collection permissions, and User groups. Make sure to thoroughly test your setup with test users. A [development instance](LINK) might come in handy.
2. If you're using any tenant groups, remove the tenant group memberships of all tenant users.
3. Change the tenant users to internal users using API:

   ```
   PUT /api/user/:id
   {"tenant_id": null}
   ```

4. Finally, disable the feature once everything is verified to work

If you don't do step 3, all your users will be deactivated.

> ## End user experience

Reviewer: I like this. Maybe we should move it up? As part of the intro/why-would-you-even-set-up-tenants section?

- **Internal collections**. Internal collections are exclusively an internal user concern. These are "normal" collections where you and other users internal to your Metabase can put the stuff that you don't want your end users to see (dashboards in development, internal metrics, maybe even analytics _about_ your tenants). Tenant users can't access any internal collections.

- **Personal collection**. Every Metabase user, including tenant users, gets a personal collection - their own space to save new questions and dashboards (of course, if they otherwise have the permissions to build and save new questions).

See [Collection permissions](#collection-permissions-for-tenants) for configuring access to different collection types.

## End user experience

End users who are members of tenants will not know that they are members of tenants.

> In the experiences that expose users to to Metabase collections (e.g. when using full-app embedding, modular embedding components with save enabled, or if tenant users are logging in directly into metabase), tenant users see all the tenant and shared collections they have access to as just collections:

Reviewer: ```suggestion
End users who are members of tenants will not know that they are members of tenants.

In the experiences that expose users to to Metabase collections (e.g. when using full-app embedding, modular embedding components with save enabled, or if tenant users are logging in directly into metabase), tenant users see all the tenant and shared collections they have access to as just collections:
```

> [SCREENSHOT]

Reviewer: Flagging this draft note.

### Special tenant slug attribute

Each tenant user will get system-defined attribute `@tenant.slug` that corresponds to the slug of the tenant. For example, if you created a tenant "Meowdern Solutions" with the slug `meowdern_solutions`, then every user from this tenant will get a special attribute `@tenant.slug : "meowdern_solutions"`.

![Slug attribute](./images/slug-attribute.png)

If you create Metabase tenants through Metabase UI, you can choose the slug when creating the tenant. If you're [using JWT to provision tenants](#provisioning-and-assigning-tenants-with-jwt), tenant slug is the value of the `@tenant` claim for JWT (or another tenant assignment attribute you selected). Slug cannot be changed later.

> The special `@tenant.slug` attribute can be used just like a normal attribute to configure attribute-based permissions like [row-level security](../permissions/row-and-column-security.md), [impersonation](../permissions/impersonation.md), or [database routing](../permissions/database-routing.md). Keep in mind that for this to work, your chosen tenant slug should correspond to how the tenant is actually identified in your setup.

Reviewer: ```suggestion
The special `@tenant.slug` attribute can be used just like a normal attribute to configure attribute-based permissions like [row-level security](../permissions/row-and-column-security.md), [impersonation](../permissions/impersonation.md), or [database routing](../permissions/database-routing.md). Your chosen tenant slug should correspond to how the tenant is actually identified in your setup.
```

### Troubleshooting JWT authentication with tenants

Some auth common error messages and what they mean:

- **Cannot add tenant claim to internal user**: JWT includes a tenant, but the user is an internal user. Only tenant users can have a tenant.
- **Tenant claim required for external user**: JWT lacks a tenant claim, but the user is an external user.
- **Tenant ID mismatch with existing user**: JWT has a different tenant than the user's assigned tenant.
- **Tenant is not active**: The tenant exists but has been deactivated.

## Data permissions for tenants

_Admin setting > Permissions_

Data permissions control what data people can see on charts and dashboards, and what they can do with that data. To control _which_ charts people see, you can use [collection permissions](#collection-permissions-for-tenants) instead.

### Data permission overview

For an overview of how data permissions work in Metabase, see [Data permissions](../permissions/data.md). Here are the highlights (but please do read the full [Data permissions](../permissions/data.md) documentation):

- **"View data"** controls what exact data each user group can see in the on on charts.

>   For example, if your tenant data is comingled in one database, then you can use [**Row and column security**](../permissions/row-and-column-security.md) or [**Impersonation**](../permissions/impersonation.md) "View data" permissions to provide tenant users with access to only certain rows and columns.

Reviewer: ```suggestion
  For example, if your tenant data is commingled in one database, then you can use [**Row and column security**](../permissions/row-and-column-security.md) or [**Impersonation**](../permissions/impersonation.md) "View data" permissions to provide tenant users with access to only certain rows and columns.
```

  If every tenant has their data in a separate database, then instead of using permissions for data access control, you can use [**database routing**](../permissions/database-routing.md) to route queries to appropriate databases directly.

> - **Create queries** controls whether tenant users can create queries on the data they see. If you want to give your tenant users the ability to drill through (e.g. through `drills` parameter in [modular embedding](../embedding/modular-embedding.md)), you need to give them "Create queries" permissions, because a drill through is a new query.

Reviewer: ```suggestion
- **Create queries** controls whether tenant users can create queries on the data they see. If you want to give your tenant users the ability to drill through (e.g., through `drills` parameter in [modular embedding](../embedding/modular-embedding.md)), you need to give them "Create queries" permissions, because a drill through is a new query.
```

### From multi-tenant to single-tenant

If you disable multi-tenant strategy, _all your tenant users will be deactivated_ and _all collections will be deleted_ (although you'll get both users and collections back if you reactivate multi-tenant strategy later). So if you want to wish to keep the active users but just disable tenancy features, you'll need some extra prep.

1. Replicate the tenant setup you have with regular groups and collections (instead of tenant groups and shared collections). Review documentation for Data permissions, Collection permissions, and User groups. Make sure to thoroughly test your setup with test users. A [development instance](../installation-and-operation/development-instance.md) might come in handy.
2. If you're using any tenant groups, remove the tenant group memberships of all tenant users.
3. Change the tenant users to internal users using API:

   ```sh
   PUT /api/user/:id
   {"tenant_id": null}
   ```

4. Finally, disable the feature once everything is verified to work

> If you don't do step 3, all your users will be deactivated.

Reviewer: We should put this in step 3.
